django/middleware/csrf.py

```diff
@@ -160,23 +161,52 @@ def _reject(self, request, reason):
         )
         return _get_failure_view()(request, reason=reason)
 
-    def process_view(self, request, callback, callback_args, callback_kwargs):
-        if getattr(request, 'csrf_processing_done', False):
-            return None
-
+    def _get_current_token(self, request):
         try:
-            cookie_token = request.COOKIES[settings.CSRF_COOKIE_NAME]
+            if settings.CSRF_USE_SESSIONS:
+                if not getattr(request, 'session', None):
+                    raise ImproperlyConfigured(
+                        'No session could be found for the request, but CSRF_USE_SETTINGS is '
+                        'turned on.'
+                    )
+                session_token = request.session[settings.CSRF_SESSION_KEY]
+                csrf_token = _sanitize_token(session_token)
+            else:
+                cookie_token = request.COOKIES[settings.CSRF_COOKIE_NAME]
+                csrf_token = _sanitize_token(cookie_token)
+                if csrf_token != cookie_token:
+                    # Cookie token needed to be replaced;
+                    # the cookie needs to be reset.
+                    request.csrf_cookie_needs_reset = True
+            # Use same token next time.
+            request.META['CSRF_COOKIE'] = csrf_token
+            return csrf_token
         except KeyError:
-            csrf_token = None
+            return None
+
+    def _set_token(self, request, response):
+        if settings.CSRF_USE_SESSIONS:
+            request.session[settings.CSRF_SESSION_KEY] = request.META['CSRF_COOKIE']
         else:
-            csrf_token = _sanitize_token(cookie_token)
-            if csrf_token != cookie_token:
-                # Cookie token needed to be replaced;
-                # the cookie needs to be reset.
-                request.csrf_cookie_needs_reset = True
-            # Use same token next time.
+            response.set_cookie(settings.CSRF_COOKIE_NAME,
+                                request.META['CSRF_COOKIE'],
+                                max_age=settings.CSRF_COOKIE_AGE,
+                                domain=settings.CSRF_COOKIE_DOMAIN,
+                                path=settings.CSRF_COOKIE_PATH,
+                                secure=settings.CSRF_COOKIE_SECURE,
+                                httponly=settings.CSRF_COOKIE_HTTPONLY
+                                )
+            # Content varies with the CSRF cookie, so set the Vary header.
+            patch_vary_headers(response, ('Cookie',))
+
+    def process_view(self, request, callback, callback_args, callback_kwargs):
+        csrf_token = self._get_current_token(request)
```

Comment:
Move that below the `csrf_processing_done` -- we do not need to do extra work in that case.